### PR TITLE
[x509-identity-mgmt] Support for external certificates

### PR DESCRIPTION
**Pull request by thiagodpf**
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

Support for accepting certificates issued by external CAs

- Trusted external CAs (CRUD)
    - Register Trusted CA Certificate
    - List Trusted CA Certificates
    - Update Trusted CA Certificate
    - Delete Trusted CA Certificate
- Register x.509 Certificate Issued by an External CA

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

no

* **Is there any issue related to this PR? (Link to an issue, e.g. #99999)** 

dojot/dojot#1540

* **Other information**:

The service has been refactored to work with IoC/DI
https://github.com/jeffijoe/awilix

### CODE REVIEW

**Comment by mprevide** [comment]
So many files =O

**Review by luissajunior — COMMENTED** [review]
For huge collections you can use estimatedDocumentCount() instead countDocuments().

**Review by luissajunior — APPROVED** [review]
Thas ALL right

File: x509-identity-mgmt/js/src/services/CertificateService.js [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+      subjectDN, this.certValidity, csrPem,
+    );
+
+    const cert = this.pkiUtils.parseCert(certificatePem);
+    const certificateFingerprint = this.pkiUtils.getFingerprint(certificatePem);
+
+    const model = new this.CertificateModel({
+      fingerprint: certificateFingerprint,
+      pem: certificatePem,
+      belongsTo,
```
Comment (rascaraficci) [inline x509-identity-mgmt/js/src/services/CertificateService.js:57]: For certificates signed by dojot, we could force `cnames` to be equal to `belongsTo`. What do you think?

File: x509-identity-mgmt/js/src/services/CertificateService.js [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+  /**
+   * Changes the ownership that is associated with a certificate
+   *
+   * @param {object} filterFields Filter fields to find the correct certificate in the database
+   * @param {object} belongsTo Data of whom the certificate should be associated
+   *
+   * @throws an exception if no record is found with the entered filters.
+   */
+  async changeOwnership(filterFields, belongsTo) {
```
Comment (rascaraficci) [inline x509-identity-mgmt/js/src/services/CertificateService.js:175]: If we consider that the cname in the CSR refers to the device to which the certificate belongs to, we need some caution here regarding certificates signed by dojot.

**Review by rascaraficci — COMMENTED** [review]
Great job. I left minor comments.

File: x509-identity-mgmt/js/config/default.conf [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+# ----------------------
+# (e. g. '/path/to/the/certificate-file.pem')
+# ----------------------
+#server.cert:string=
+
+# ---------------------
+# X509IDMGMT_SERVER_KEY
+# ---------------------
+# (e.g. '/path/to/the/private.key')
+#server.key:string=
```
[the remark this answers is not in the document]
Comment (thiagodpf) [reply x509-identity-mgmt/js/config/default.conf:23]: done

File: x509-identity-mgmt/js/src/core/schemaValidator.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ],
+  });
+
+  /**
+ * Format error responses
+ * @param  {String} schema$id - ID of the schema used to validate json
+ * @param  {Object} schemaErrors - array of json-schema errors, describing each validation failure
+ * @return {String} formatted api response
+ */
+  function errorResponse(schema$id, schemaErrors) {
```
[the remark this answers is not in the document]
Comment (thiagodpf) [reply x509-identity-mgmt/js/src/core/schemaValidator.js:25]: I must have followed some doc example and forgot to change

File: x509-identity-mgmt/js/src/services/CertificateService.js (hunk 1 above)
Comment (thiagodpf) [reply x509-identity-mgmt/js/src/services/CertificateService.js:57]: I think that all CSR `Subject DN` fields could be ignored and dojot could generate a UUIDv4 as the value of de `cname` field in the issued certificate.

File: x509-identity-mgmt/js/src/services/CertificateService.js (hunk 2 above)
Comment (thiagodpf) [reply x509-identity-mgmt/js/src/services/CertificateService.js:175]: With de IoTAgent relying on the certificate's fingerprint, we no longer have to worry about any value of the `Subject DN` fields. I think the association should be only in the `belongsTo` attribute of the request's payload.

File: x509-identity-mgmt/js/src/services/CertificateService.js (hunk 1 above)
Comment (rascaraficci) [reply x509-identity-mgmt/js/src/services/CertificateService.js:57]: I agree with you!

File: x509-identity-mgmt/js/src/services/CertificateService.js (hunk 2 above)
Comment (rascaraficci) [reply x509-identity-mgmt/js/src/services/CertificateService.js:175]: OK!

File: x509-identity-mgmt/js/src/decorators/decoration.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+   * problem to be solved, but the fundamental point is the centralization of this
+   * logic, thus avoiding duplication of code.
+   * @param {Object} target where decorators will be kept.
+   * @param {Array} methods that will be decorated.
+   * @param {Array} decorators decorators with the logic to be applied on the original methods.
+   */
+  function decorate(target, methods, decorators) {
+    // The list must be inverted so that decorators are executed
+    // in the correct  order when the decorated method is called
+    decorators.reverse();
```
[the remark this answers is not in the document]
Comment (thiagodpf) [reply x509-identity-mgmt/js/src/decorators/decoration.js:14]: I improved the comment to be clearer

File: x509-identity-mgmt/js/src/decorators/decoration.js [hunk 6]
```diff
@@ -9,8 +9,12 @@ module.exports = ({ levelDebug }) => {
...
    */
   function decorate(target, methods, decorators) {
-    // The list must be inverted so that decorators are executed
-    // in the correct  order when the decorated method is called
+    // For the order of execution of the decorators to follow the same order
+    // in which they were declared in the list, it is necessary to apply the
+    // decoration starting with the last decorator of the list and thus until
+    // the first...
+    // Instead of going through the list from the end to the beginning, we can
+    // invert it and then iterate through it with the "forEach()" method.
```
Comment (rascaraficci) [inline x509-identity-mgmt/js/src/decorators/decoration.js:17]: :+1:

**Review by rascaraficci — APPROVED** [review]
Good job!

**Review by luissajunior — APPROVED** [review]
Approved

**Comment by mprevide** [comment]
As this pr will be not part of release v0.6.0, maybe it should be merged into a feature branch. 
I don't know ..

**Pull request merged**